### ISSUE TITLE
IDE0016 codefix can change behaviour

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/roslyn#36177]
**Version Used**: 16.1.1+28922.388

**Steps to Reproduce**:

```csharp
object _o;
bool _b;
void M(object o)
{
    if (o == null)
        throw new ArgumentNullException(nameof(o));
    _b = true;
    _o = o;
}
```

IDE0016 fires and wants to change the code to:

```csharp
object _o;
bool _b;
void M(object o)
{
    _b = true;
    _o = o ?? throw new ArgumentNullException(nameof(o));
}
```

**Expected Behavior**:

No change in behaviour.

**Actual Behavior**:

Potential change in behaviour due to throw being moved after side effecty code.

**Comment by CyrusNajmabadi** [comment]
Yes.  This was a known limitation when the feature was being created.  The major problem was that if we limited to only perfectly safe scenarios, it basically reduced the useful this to the point of barely being offerable at all.  We found, in practice, that in nearly all the cases it was offered, even if it couldn't be proven safe, it would be safe.  So we went with allowing it.  Note that this is the bar for all local fixes.  Nearly all of them have some type of input they may break, but by and large there is enough value in being there in 99%+ of cases when it turns out to be safe, versus not having it appear there.

**Comment by drewnoakes** [comment]
@CyrusNajmabadi did the team consider using the current behaviour only in constructors where side effects to the current instance tend to be redundant if an exception is raised, and being more conservative in non-constructors?

In my experience the majority of such throw expression transformations are in constructors, mapping arguments to fields. Though of course there are still constructor cases where deferring throws could be incorrect.

**Comment by CyrusNajmabadi** [comment]
I did a pass over a few large codebases (include corefx) to see how often i found arg checking and some of statefulness, before finally assigning the arg to something.  I think it was on the order of 1000:1 or more of that happening**.  So i basically didn't think it was relevant enough.  

** I think the actual number was that we found 2 cases in all the code checked.  So it was felt it was rare enough to be acceptable.

I think it might always be interesting to reconsider, if more feedback comes in about this though.

**Comment by stephentoub** [comment]
> I think it was on the order of 1000:1 or more of that happening

This rule has resulted in a significant number of breaks for me, to the point where I've just stopped trying to use it. I just ran it on Corelib and counted the number of times it was correct and the number of times it changed program behavior.  Across 42 files, it was correct with 53 transformations, and incorrect with 26, meaning if I didn't carefully scrutinize these changes, I would have checked in 26 bugs.  That's a 30% failure rate for the rule.  That's way, way, way too high.

**Comment by CyrusNajmabadi** [comment]
Note: one thing that came from a discussion on this was that there's a difference between constructors versus normal methods in terms of side-effects and chance of being able to observe an issue.  

IMO, we likely need to tweak things to be conservative if this a non-constructor scenario.  I think my reporting on 1000:1 is incorrect and the analysis was done just when thi was a constructor feature.  This is my bad and I'm sorry due diligence wasn't done as this feature morphed from the initial design.

**Comment by CyrusNajmabadi** [comment]
@stephentoub in your analysis, how many of those cases were normal methods and how many were constructors?  Thanks!

**Comment by stephentoub** [comment]
> How many of those cases were normal methods and how many were constructors? Thanks!

I'd have to go back and re-count, but a bunch were ctors IIRC.  Note that this isn't just about state mutation, which is I expect more common in non-constructors.  It's also about, for example, exceptions that end up changing as a result.

For example, take BinaryWriter's ctor:
https://github.com/dotnet/coreclr/blob/e1fa3c83171e09ec29481c82c3042c30146fc4d4/src/System.Private.CoreLib/shared/System/IO/BinaryWriter.cs#L51-L65
```C#
        public BinaryWriter(Stream output, Encoding encoding, bool leaveOpen)
        {
            if (output == null)
                throw new ArgumentNullException(nameof(output));
            if (encoding == null)
                throw new ArgumentNullException(nameof(encoding));
            if (!output.CanWrite)
                throw new ArgumentException(SR.Argument_StreamNotWritable);

            OutStream = output;
            _buffer = new byte[16];
            _encoding = encoding;
            _encoder = _encoding.GetEncoder();
            _leaveOpen = leaveOpen;
        }
```
When you fix all IDE0016, it becomes this:
```C#
        public BinaryWriter(Stream output, Encoding encoding, bool leaveOpen)
        {
            if (output == null)
                throw new ArgumentNullException(nameof(output));
            if (!output.CanWrite)
                throw new ArgumentException(SR.Argument_StreamNotWritable);

            OutStream = output;
            _buffer = new byte[16];
            _encoding = encoding ?? throw new ArgumentNullException(nameof(encoding));
            _encoder = _encoding.GetEncoder();
            _leaveOpen = leaveOpen;
        }
```
If this change were applied and someone were calling `new BinaryWriter(nonWritableStream, null, true)`, this would now throw an `ArgumentException` instead of an `ArgumentNullException`, because it's switching the order in which the checks are applied.  Now, that might be a change we decide is ok to take if the reward was worthwhile, but it's not for such minimal gains as this refactoring yields.  This applies to non-constructors as well, for example here's NumberFormatInfo.CurrencySymbol:
https://github.com/dotnet/coreclr/blob/98bf56a2cf19645563b41ab01f6972cbe83c9d44/src/System.Private.CoreLib/shared/System/Globalization/NumberFormatInfo.cs#L361-L374
```C#
        public string CurrencySymbol
        {
            get => _currencySymbol;
            set
            {
                if (value == null)
                {
                    throw new ArgumentNullException(nameof(value));
                }

                VerifyWritable();
                _currencySymbol = value;
            }
        }
```
After applying IDE0016's fix all, it looks like this:
```C#
        public string CurrencySymbol
        {
            get => _currencySymbol;
            set
            {
                VerifyWritable();
                _currencySymbol = value ?? throw new ArgumentNullException(nameof(value));
            }
        }
```
So now if the value is null and the instance is ReadOnly, the caller will get an InvalidOperationException instead of an ArgumentNullException.

Some cases are potentially more devious and do involve shared state, of course.  For example, consider DiagnosticCounter's ctor:
https://github.com/dotnet/coreclr/blob/98bf56a2cf19645563b41ab01f6972cbe83c9d44/src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/DiagnosticCounter.cs#L33
```C#
        internal DiagnosticCounter(string name, EventSource eventSource)
        {
            if (name == null)
            {
                throw new ArgumentNullException(nameof(Name));
            }

            if (eventSource == null)
            {
                throw new ArgumentNullException(nameof(EventSource));
            }

            _group = CounterGroup.GetCounterGroup(eventSource);
            _group.Add(this);
            Name = name;
            DisplayUnits = string.Empty;
            EventSource = eventSource;
        }
```
And after the fix all is applied:
```C#
        internal DiagnosticCounter(string name, EventSource eventSource)
        {
            if (eventSource == null)
            {
                throw new ArgumentNullException(nameof(EventSource));
            }

            _group = CounterGroup.GetCounterGroup(eventSource);
            _group.Add(this);
            Name = name ?? throw new ArgumentNullException(nameof(Name));
            DisplayUnits = string.Empty;
            EventSource = eventSource;
        }
```
Here GetCounterGroup gets a process-wide share group instance and then adds this instance to it... that's fine when there's no legitimate chance for an exception after adding this instance, but this transformation is actually doing the argument validation after adding the instance and before fully setting up the instance, so now the group contains an instance that will remain in an incomplete state.

**Comment by CyrusNajmabadi** [comment]
Yup.  Those are cases that came up, with the understanding htat might be a problem.  But this is not to diminish your concerns or frustration here.  I def think there needs to be a tweak here.  We may have to really go for more of an approach that we only move these checks past code that seems much safer.  

Here's the pity of it.  We found a ton of cases like this:

```c#
public Whatever(string arg1, object arg2, bool etc)
{
    if (arg1 == null)
    {
        throw ...
    }
    OurCustomArgChecker(arg2); // note: this was fairly common corefx
    this.arg1 = arg1;
}
```

We want to be able to say "if there's a block of arg checking, we likely have high enough confidence to move the check, even if that reorders".  However, defining "a block of arg checking" isn't easy to do (note: that doesn't excuse anything).  We can totally look for common patters.  i.e. "generally an `if+throw`" or more `x = x ?? throw` statements.  But if we are conservative, then cases like the above will cause us to not offer anymore.

Based on your feedback though, i'm thinking:
1. we need to bite the bullet and make that change.
2. we can add a warning, and at least *not* have fix-all update these.

I'm leaning toward the latter option. I like the idea of at least still letting people know about these cases, but making it so that they have to explicitly choose to use the suggestion here, and accept the warning on a case by case basis so they think about the implication of the reorder.

**Comment by stephentoub** [comment]
> I like the idea of at least still letting people know about these cases, but making it so that they have to explicitly choose to use the suggestion here, and accept the warning on a case by case basis so they think about the implication of the reorder.

Me, too.  It's another case where the "confidence" indication from the diagnostic/refactoring would be useful; a developer could choose to start by applying to all the high-confidence cases, and then explore from there cases where there's lower confidence but still potentially refactorings of interest.

Thanks for listening.

**Comment by CyrusNajmabadi** [comment]
Note for IDE team, i'm willing to contribute fixes here.

**Comment by CyrusNajmabadi** [comment]
>Thanks for listening.

Least I can do.  Thanks for all the time and effort to bring this up, really make the issue clear with real world cases, and be willing to discuss strategies that can improve things without having to throw the baby out :)  

Please keep filing the cases (with other IDE fixes, since i think this one is clear), *especially* around teh 'confidence' angle.  I've seen all the stuff you've filed that is just straight "bugs" (usually around comment trivia), but the "confidence" ones are hte ones that def require some really deep examination and thought, so if you can bring those to the team's (and my) attention, that would be incredibly valuable.   

The deep belief is that these features are never truly "done", but often attention is only really given to them based on feedback.  So by putting the spotlight on these features, it really does *directly* drive to improvements for everyone.  

Thanks!

**Comment by sharwell** [comment]
Design review conclusion: We still believe this will be largely addressed by C# 9 having new syntax to trigger automatic argument null checking. In the meantime, we can avoid problems for the majority of users by simply changing the default severity of this diagnostic from **suggestion** to **hidden**.

**Comment by glen-84** [comment]
Duplicate of #22325.
